### PR TITLE
Simplify ViewViewModelFactory

### LINKED ISSUE #1361: Creating pull request fails with: "Request failed with status code: 403"
- GitHub Extension for Visual Studio version: 2.3.7 (dev)
- Visual Studio version: 2015

1. Clicked on `Pull Requests`, `Create New`:

image

2. Entered title and description:

image

3. Request fails with `Request failed with status code: 403`.

Same happened after branch was created and pushed.

### LINKED ISSUE #1368: Simplify the way ViewViewModelFactory uses MEF
- GitHub Extension for Visual Studio version: 2.3.7 (dev)

The `ViewViewModelFactory` class currently has dependencies on `ICompositionService` and `IGitHubServiceProvider`, which I'm not sure are necessary. It also uses `AllowRecomposition = true`, which I don't think is supported in Visual Studio's flavor of MEF.

I'd like to see if this can be simplified, so we don't confuse people looking at this class in future.

### PR DESCRIPTION
**Pull request by jcansdale**
This PR:

- Removes unnecessary call to `SatisfyImportsOnce`
- Removed unsupported setting of `AllowRecomposition`
- Stores `ViewModelType` as a `string` in metadata to avoid pre-loading assemblies

Storing the `ViewModelType` as a `string` mirrors the way `Export` metadata is stored. For example, the export: `Export(typeof(Foo.IBar))` will create a metadata property `ExportTypeIdentity` with the string "Foo.IBar".

Removing the dependency on `IGitHubServiceProvider` is more complicated and might be tackled in a separate PR, see: 
https://github.com/github/VisualStudio/pull/1370#discussion_r156337188

Fixes #1361

### CODE REVIEW

File: src/GitHub.Exports/Exports/ExportMetadata.cs [hunk 1]
```diff
@@ -69,7 +69,7 @@ public ExportMenuAttribute() : base(typeof(IMenuHandler))
     public interface IViewModelMetadata
     {
         [SuppressMessage("Microsoft.Performance", "CA1819:PropertiesShouldNotReturnArrays")]
-        Type[] ViewModelType { get; }
+        string[] ViewModelType { get; }
```
Comment (grokys) [inline src/GitHub.Exports/Exports/ExportMetadata.cs]: Why did you change the type of this member? It's not mentioned in the PR description.

Comment (jcansdale) [reply src/GitHub.Exports/Exports/ExportMetadata.cs]: My aim was to remove `ViewViewModelFactory`s dependency on `IGitHubServiceProvider`.

Instead of calling `IGitHubServiceProvider.ExportProvider.GetExport<TViewModel>()`, I was hoping to use a second `ExportFactory`, like so:

```
        [ImportMany]
        IEnumerable<ExportFactory<FrameworkElement, IViewModelMetadata>> Views { get; set; }

        [ImportMany("ViewModel")]
        IEnumerable<ExportFactory<object, IViewModelMetadata>> ViewModels { get; set; }
```

We would then enumerate through the `IViewModelMetadata` when finding a factory to create a View or ViewModel.  If the `IViewModelMetadata` contained a `Type` rather than a `string`, it would force *all* assemblies to load the first time a View or ViewModel was created (hence the change to using `string`).

Unfortunately there appear to be some strange limitations with `ExportFactory` which prevent this from working. 😢

Comment (jcansdale) [reply src/GitHub.Exports/Exports/ExportMetadata.cs]: I think the change from `Type` to `string` is still valuable because it should prevent all assemblies that contain a view from being pre-loaded when the first view is created.  Does that make sense?

**Comment by jcansdale** [comment]
Note to self, might need to use something like this:

```
        [Export]
        public class RootPart
        {
            [Import, SharingBoundary("SomeBoundary")]
            public ExportFactory<BoundaryPart> Factory { get; set; }
        }

      [Export, Shared("SomeBoundary")]
```

File: src/GitHub.Exports/Exports/ExportMetadata.cs (hunk 1 above)
Comment (grokys) [reply src/GitHub.Exports/Exports/ExportMetadata.cs]: Yes, that makes sense, need to check that this is actually happening!
